src/systems/filecoin_vm/interpreter/vm_interpreter.go

```diff
@@ -20,177 +21,118 @@ func (vmi *VMInterpreter_I) ApplyMessageBatch(inTree st.StateTree, msgs []Messag
-	compTree, toActor, err = treeGetOrCreateAccountActor(compTree, message.To())
-	if err != nil {
-		return applyError(inTree, exitcode.ActorNotFound, gascost.ApplyMessageFail)
+	// TODO: Should we allow messages to implicitly create actors?
+	// - If so, need to specify how the actor code is determined
+	// - We should be able to disallow this and still preserve functionality by
+	//   requiring two separate messages (create actor, transfer initial funds)
+	createRet := compTree.WithNewActorIfMissing(message.To())
+	if createRet.Which() == st.StateTree_WithNewActorIfMissing_FunRet_Case_error {
+		return inTree, _applyError(exitcode.ActorNotFound)
+	} else {
+		compTree = createRet.As_StateTree()
 	}
 
-	// deduct maximum expenditure gas funds first
-	// TODO: use a single "transfer"
-	compTree = treeDeductFunds(compTree, fromActor, maxGasCost)
-
-	// transfer funds fromActor -> toActor
-	// (yes deductions can be combined, spelled out here for clarity)
-	// TODO: use a single "transfer"
-	compTree = treeDeductFunds(compTree, fromActor, message.Value())
-	compTree = treeDepositFunds(compTree, toActor, message.Value())
-
-	// Prepare invocInput.
-	invocInput := vmr.InvocInput{
-		InTree:    compTree,
-		FromActor: fromActor,
-		ToActor:   toActor,
-		Method:    message.Method(),
-		Params:    message.Params(),
-		Value:     message.Value(),
-		GasLimit:  message.GasLimit(),
+	toActor := compTree.GetActor(message.To())
+	if toActor == nil {
+		panic("Interpreter error: actor present (or created) but not retrieved")
 	}
-	// TODO: this is mega jank. need to rework invocationInput + runtime boundaries.
-	invocInput.Runtime = makeRuntime(compTree, invocInput)
-
-	// perform the method call to the actor
-	// TODO: eval if we should lift gas tracking and calc to the beginning of invocation
-	// (ie, include account creation, gas accounting itself)
-	out := invocationMethodDispatch(invocInput)
 
-	// var outTree StateTree
-	if out.ExitCode != 0 {
-		// error -- revert all state changes -- ie drop updates. burn used gas.
-		outTree = inTree // wipe!
-		outTree = treeDeductFunds(outTree, fromActor, gasToFIL(out.GasUsed, message.GasPrice()))
-
-	} else {
+	// deduct maximum expenditure gas funds first
+	compTree = _withTransferFundsAssert(compTree, message.From(), vmr.BurntFundsActorAddr, maxGasCost)
+
+	rt := vmr.Runtime_Make(
+		compTree,
+		message.From(),
+		actor.TokenAmount(0),
+		message.GasLimit(),
+	)
+
+	sendRet, sendRetStateTree := rt.Impl().SendToplevelFromInterpreter(
+		message.To(),
+		msg.InvocInput_Make(
+			message.Method(),
+			message.Params(),
+			message.Value(),
+		),
+		false,
+	)
+
+	if sendRet.ExitCode().AllowsStateUpdate() {
 		// success -- refund unused gas
-		outTree = out.OutTree // take the state from the invocation output
-		refundGas := message.GasLimit() - out.GasUsed
-		outTree = treeDepositFunds(outTree, fromActor, gasToFIL(refundGas, message.GasPrice()))
-		outTree = treeIncrementActorSeqNo(outTree, fromActor)
```

Comment:
- I do think there is utility in keeping a counter for message invocations, so that (for example) an actor would be able to get the number: `(epoch, callNum)` where `callNum` increments with each internal send in an epoch 
  - this would be useful for randomness generation (eg method can ask for `rt.Randomness(rt.Epoch(), rt.CallNum())`
- potentially even `(epoch, extCallNum, intCallNum)` where `extCallNum` is the external call number (eg the root invocation -- the message), and `intCallNum` is the internal call number in that stack.

Reply:
Sounds good -- I've modified the code to use the existing `CallSeqNum` counter only for external calls. Should we add the internal call counter also, or postpone that change for now?

Reply:
> Should we add the internal call counter also, or postpone that change for now?

maybe wait until we need it? (one place i thought we needed it was the `ComputeActorAddress`, but i think `CallSeqNum` is enough)